src/validation.cpp

```diff
@@ -1288,10 +1311,54 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptMultipleTransactions(const std::
                    [](const auto& tx) { return Workspace(tx); });
     std::map<uint256, MempoolAcceptResult> results;
 
+    // Not a complete check of v3 rules, helps us detect failures early.
+    // Some data structures are only necessary if we are checking v3 rules; skip this work if those
+    // rules are not applicable.
+    const bool check_v3_rules{std::any_of(txns.cbegin(), txns.cend(), [](const auto& tx){ return tx->nVersion == 3; })};
+    if (check_v3_rules) {
+        if (const auto err_string{PackageV3SanityChecks(txns)}) {
+            package_state.Invalid(PackageValidationResult::PCKG_POLICY, "v3-violation", err_string.value());
+            return PackageMempoolAcceptResult(package_state, std::move(results));
+        }
+    }
+
+    // Map from txid of a transaction to its in-package ancestor set. Not MemPoolAccept-wide
+    // because ancestor sets may change between multiple calls to AcceptMultipleTransactions, i.e.
+    // due to something being accepted.
+    const auto in_package_ancestors{CalculateInPackageAncestors(txns)};
+
     LOCK(m_pool.cs);
 
     // Do all PreChecks first and fail fast to avoid running expensive script checks when unnecessary.
     for (Workspace& ws : workspaces) {
+        // This process is O(n^2) in the number of transactions, so only do it when v3 rules are
```

Comment:
this block is worth encapsulating and adding unit tests to ensure the populated sets `ws.m_num_in_package_ancestors` and `ws.m_ancestors_of_in_package_ancestors` match expected

Reply:
Thank you for your interest, my dear friend. Can you explain further? Thank you

Reply:
Not 100% sure how to encapsulate this exactly since Workspace is private to MemPoolAccept, which is private to validation. We have some tests that the ancestor sets are correctly calculated. Maybe we add a helper function? Or I can add some more assumes.